Userland/Libraries/LibDiff/Generator.cpp

```diff
@@ -63,21 +63,43 @@ Vector<Hunk> from_text(StringView const& old_text, StringView const& new_text)
     }
 
     Vector<Hunk> hunks;
+    Hunk cur_hunk;
+    bool in_hunk = false;
+
+    auto update_hunk = [&](size_t i, size_t j, Direction direction) {
+        if (!in_hunk) {
+            in_hunk = true;
+            cur_hunk = { i, j, {}, {} };
+        }
+        if (direction == Direction::Down) {
+            cur_hunk.added_lines.append(new_lines[j]);
+        } else if (direction == Direction::Right) {
+            cur_hunk.removed_lines.append(old_lines[i]);
+        }
+    };
+
     size_t i = 0;
     size_t j = 0;
 
-    // FIXME: This creates a hunk per line, very inefficient.
     while (i < old_lines.size() && j < new_lines.size()) {
         auto& cell = dp(i, j);
         if (cell.direction == Direction::Down) {
-            hunks.append({ i, j, {}, { new_lines[j] } });
+            update_hunk(i, j, cell.direction);
             ++j;
         } else if (cell.direction == Direction::Right) {
-            hunks.append({ i, j, { old_lines[i] }, {} });
+            update_hunk(i, j, cell.direction);
             ++i;
         } else {
             ++i;
             ++j;
+            if (in_hunk) {
+                if (cur_hunk.added_lines.size() > 0)
+                    cur_hunk.target_start_line++;
+                if (cur_hunk.removed_lines.size() > 0)
+                    cur_hunk.original_start_line++;
+                hunks.append(cur_hunk);
+                in_hunk = false;
+            }
         }
     }
 
```

Comment:
Off-by-one error: If `in_hunk = true` in the end, then that hunk is lost.
Suggestion: Create a `flush_hunk` that contains the entire `if (in_hunk) { … }` logic, and call it both in the `else` branch, as well as after the `while` loop.

Also, to make sure that this type of error never crops up again: How do you feel about writing a few tests? I strongly recommending *first* writing a test case that shows that the current code is broken, and only *then* fixing the code. Of course, the commits should not reflect that.

Reply:
Adding tests is definitely a good idea, though I don't have much time these days to sit down and write a bunch of them. I'll leave for a future PR when I get some free time!